system_metrics_collector/src/topic_statistics_collector/subscriber_topic_statistics.cpp

```diff
@@ -0,0 +1,271 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <limits>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "rcl/time.h"
+#include "rclcpp/rclcpp.hpp"
+#include "rcpputils/asserts.hpp"
+
+#include "received_message_age.hpp"
+#include "received_message_period.hpp"
+#include "subscriber_topic_statistics.hpp"
+#include "system_metrics_collector/constants.hpp"
+
+using metrics_statistics_msgs::msg::MetricsMessage;
+
+namespace topic_statistics_collector
+{
+static rcl_interfaces::msg::IntegerRange buildPositiveIntegerRange(
+  int64_t from,
+  int64_t to,
+  uint64_t step)
+{
+  rcl_interfaces::msg::IntegerRange positive_range;
+  positive_range.from_value = from;
+  positive_range.to_value = to;
+  positive_range.step = step;
+
+  return positive_range;
+}
+
+static rcl_interfaces::msg::ParameterDescriptor buildPeriodParameterDescriptor(
+  const std::string & description)
+{
+  const auto positive_range = buildPositiveIntegerRange(
+    1,
+    std::numeric_limits<decltype(rcl_interfaces::msg::IntegerRange::to_value)>::max(),
+    1
+  );
+
+  rcl_interfaces::msg::ParameterDescriptor period_descriptor;
+  period_descriptor.read_only = true;
+  period_descriptor.integer_range.push_back(positive_range);
+  period_descriptor.description = description;
+
+  return period_descriptor;
+}
+
+static rcl_interfaces::msg::ParameterDescriptor buildTopicParameterDescriptor(
+  const std::string & description)
+{
+  rcl_interfaces::msg::ParameterDescriptor topic_descriptor;
+  topic_descriptor.read_only = true;
+
+  return topic_descriptor;
+}
+
+static void validateStringParam(const std::string & param)
+{
+  if (param.empty()) {
+    std::stringstream ss;
+    ss << param << " node paramater cannot be empty";
+    throw std::invalid_argument{ss.str().c_str()};
+  }
+}
+
+static void initializeClock(rcl_clock_t & clock_)
+{
+  auto allocator = rcl_get_default_allocator();
+  if (RCL_RET_OK != rcl_clock_init(RCL_STEADY_TIME, &clock_, &allocator)) {
+    throw std::runtime_error{"Failed to initialize clock"};
+  }
+}
+
+static rcl_time_point_value_t getCurrentTime(rcl_clock_t & clock)
+{
+  rcl_time_point_value_t now_time_point;
+  if (RCL_RET_OK != rcl_clock_get_now(&clock, &now_time_point)) {
+    RCUTILS_LOG_ERROR_NAMED("SubscriberTopicStatistics", "Failed to get current time");
+    return 0;
```

Comment:
Is a value of 0 handled elsewhere or is this documented?

Reply:
It is not handled in this module. Individual collectors are expected to handle it as required.